Jetpack Boost: Update data sync flexibility

pyronaur opened the pull request:
During the meetup we discovered some parts of the data-sync package that proved to be too rigid to fully embrace data-sync.

This PR addresses the flexibility issues allowing each data-sync entry to control which endpoints it creates and how the data is handled. In addition, this allows creating read-only stores, which has been the top-requested feature by @haqadn for a while now.

This is still WIP, so the documentation needs to be updated, but here are some snippets in the meantime:

#### Use the Merge endpoint:

```ts
const partial_example = client.createAsyncStore(
	'partial_example',
	z.object( {
		unchanging_timestamp: z.number(),
		incrementing_number: z.number(),
	} )
);

partial_example.setSyncAction( async (prevValue, value, abortSignal) => {
	console.log("Previous value:", prevValue.incrementing_number);
	console.log("New value:", value.incrementing_number);
	return await partial_example.endpoint.MERGE( {
		incrementing_number: value.incrementing_number,
	}, abortSignal );
} );

setInterval( () => {
	partial_example.store.update( value => {
		value.incrementing_number++;
		return value;
	} );
}, 2000 );

partial_example.store.subscribe( value => {
	console.log( 'partial_example', value );
} );
```

#### Registering a new option using Schema

```php
$partial_example = Schema::as_assoc_array(
	array(
		'unchanging_timestamp' => Schema::as_number(),
		'incrementing_number'  => Schema::as_number(),
	)
)->fallback(
	array(
		'unchanging_timestamp' => time(),
		'incrementing_number'  => 0,
	)
);

class Partial_Example implements Entry_Can_Get, Entry_Can_Set, Entry_Can_Merge {
	public function get() {
		return get_option( 'partial_example' );
	}

	public function merge( $value, $partial_value ) {
		$value['incrementing_number'] = $partial_value['incrementing_number'];
		return $value;
	}

	public function set( $value ) {
		update_option( 'partial_example', $value );
	}
}

jetpack_boost_register_option( 'partial_example', $partial_example, new Partial_Example() );
```

### Other information:
- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
N/A

## Does this pull request change what data or activity we track or use?
N/A

## Testing instructions:
🦄

pyronaur committed:
Replace Data_Sync_Entry and Storage with Schema

pyronaur committed:
Refactor into a Data_Sync_Entry wrapper

pyronaur committed:
Improve endpoint safety and consistency

pyronaur committed:
Update endpoints to match WP JS DataSync updates

pyronaur committed:
Only allow GET and POST on the "get" endpoint

pyronaur committed:
Require "Entry_Can_Set" to be present for Merge

pyronaur committed:
Wrap all entries with Data_Sync_Entry

pyronaur committed:
Remove unused T

pyronaur committed:
Pass the previous, schema validated value to merge

pyronaur committed:
Partial type doesn't work at the moment

pyronaur committed:
Pass the previous value to SyncAction

pyronaur committed:
Clone the previous value to break the reference

pyronaur committed:
Update changelogs

haqadn commented on projects/plugins/boost/wp-js-data-sync.php:
```diff
@@ -1,22 +1,26 @@
+	}
 	return Registry::get_instance( JETPACK_BOOST_DATASYNC_NAMESPACE )
-					->register( $name, $handler );
+					->register( $key, new Data_Sync_Entry( $entry, $schema ) );
```
Since the reason behind `Data_Sync_Entry` is to enforce the schema validation, and per my understanding it is not supposed to be customized, shouldn't this be an internal class, and should the registry have an interface like:

```
public function register( $key, $schema, $entry )
```

And then, the registry instantiates the `Data_Sync_Entry` class as a wrapper of the $entry(an instance of `Data_Sync_Option` by default).

Also, it looks like the namespace needs to be provided in two places. Can we require the storage handlers to have a method like `setNamespace`, which the registry then uses to set the namespace while registering?

pyronaur replied:
> Also, it looks like the namespace needs to be injected in two places. Can we require the storage handlers to have a method like setNamespace, which the registry then uses to set the namespace while registering?

I think it's good to namespace options, which is why Data_Sync_Option requires a namespace upon instantiation.  However, not every Entry type requires a namespace, like in the example given in the PR description.

The code here might look like one namespace depends on the other, but that's not the case. I could, and probably should, and probably will, rename the option namespace to option prefix - it can be the same thing as the actual endpoint namespace, but the option namespace isn't related to it in any other way than just by name.

---

Side note: 

The reason for Data_Sync_Entry is not only to enforce a schema, but to also provide a common and predictable entry interface. 

---

As for moving more of the logic inside the register method - I think that's not the right approach - it would make "the registry" do more than just "register a reference" - it would also be responsible for setting up.

But your prodding made me realize here that Data_Sync_Entry as the argument is also incorrect because it requires a concrete class and not a contract. I'll make a few changes there to ensure that part is flexible and clear too.

pyronaur committed:
Remove code that's no longer used

pyronaur committed:
Introduce a Data_Sync_Entry_Adapter

pyronaur committed:
Simplify Data_Sync_Option

pyronaur committed:
Improve Boolean default value handling

haqadn commented on projects/js-packages/svelte-data-sync-client/src/SyncedStore.ts:
```diff
@@ -68,22 +69,22 @@ export class SyncedStore< T > {
-			const result = await this.updateCallback( value, this.abortController.signal );
+			const result = await this.syncAction( prevValue, value, this.abortController.signal );
 
 			// Success is only when the updateCallback result matches the value.
 			if ( this.equals( result, value ) ) {
```
This creates weird issues when the value changes in the backend.

i.e. When you are changing module status in multiple tabs, it will condemn one of them to fail.

pyronaur replied:
Fixed

pyronaur committed:
Refactor Data_Sync setup

haqadn commented on projects/js-packages/svelte-data-sync-client/src/initializeClient.ts:
```diff
@@ -120,8 +120,8 @@ export function initializeClient( namespace: string ) {
 		 *	const client = initializeClient( 'jetpack_favorites' );
-		 *	client.setCallback( 'status', ( value ) => {
-		 *		client.endpoint.status.POST( value.replace("a", "b") );
+		 *	client.setSyncAction( 'status', ( value ) => {
+		 *		client.status.endpoint.SET( value.replace("a", "b") );
```
```suggestion
		 *	client.setSyncAction( ( preValue, value, abortSignal ) => {
		 *		return client.status.endpoint.SET( value, abortSignal );
```

haqadn commented on projects/packages/wp-js-data-sync/src/class-data-sync.php:
```diff
@@ -134,27 +136,21 @@ public function add_to_plugin_page( $plugin_page, $parent_page ) {
 
-		return $instance;
+	public function register( $key, $schema, $entry = null ) {
+		if ( ! $entry instanceof Data_Sync_Entry_Adapter ) {
```
`Data_Sync_Entry` is an instance of `Data_Sync_Entry_Adapter` and entry is just supposed to be an instance of `Entry_Can_Get` based on what I understand. Is this check wrong?

pyronaur committed:
Add typehint

pyronaur committed:
Add a test for nested fallbacks

pyronaur committed:
Test passing a partial array

pyronaur committed:
Update projects/js-packages/svelte-data-sync-client/src/initializeClient.ts

Co-authored-by: Adnan Haque <3737780+haqadn@users.noreply.github.com>

pyronaur committed:
Remove object equality checking

Instead - accept whatever the server provides, as long as it's a valid Zod shape

pyronaur committed:
Registering an entry requires Entry_Can_Get

pyronaur committed:
Add a docblock

haqadn reviewed (dismissed):
LGTM!

pyronaur committed:
Fix unused variable

pyronaur committed:
Error logging is ok here

haqadn reviewed (approved):
Ready to merge
